### DISCUSSION TITLE
[Extension Proposal] QT-EWMA Concept-Drift Detection Algorithm

### DISCUSSION THREAD

**Discussion by DiegoStucchi** [online-ml/river#1199] [Ideas]
Hello River Team, my name is Diego Stucchi, and I am a Ph.D. student from Politecnico di Milano. My main research topic is change and concept-drift detection. My colleagues and I worked on developing the QuantTree [1] algorithm and its extensions, including QT-EWMA [2], a concept-drift detector that controls the Average Run Length. We have a public version of our algorithms [3] published on GitHub, and we would be interested in contributing to river. What are the requirements to join this project? Are there restrictions on the license? Is there a guide for (potential) contributors?

[1] [QuantTree: Histograms for Change Detection in Multivariate Data Streams](https://proceedings.mlr.press/v80/boracchi18a.html)
[2] [QT-EWMA: QuantTree for change detection in multivariate datastreams](https://boracchi.faculty.polimi.it/Projects/QT-EWMA.html)
[3] [GitHub Repository](https://github.com/diegocarrera89/quantTree)

**Comment by MaxHalford** [comment]
Hello there, thanks for getting in touch 👋

> My main research topic is change and concept-drift detection

Would you mind giving us a brief explanation of the difference between the two? I'm asking because I'm wondering if we need to make a distinction in our API interface. Right now, we have a `drift` module. Some Polytechnique students have coded [this](https://github.com/wimmerth/onlinelearning-changepoints/), where they introduce a new base class for change-point detection. My gut feeling is that this isn't fundamentally different to drift detection.

> QuantTree [1] algorithm

Looks interesting! We already have a `sketch.Histogram` class. Maybe you can reuse it? However it's univariate, and your method seems to require multivariate histograms, correct?

I don't see any mention of throughput in the paper. Do you have some rough complexity/benchmark at hand?

> QT-EWMA [2], a concept-drift detector that controls the Average Run Length

Is this always better than QuantTree? I'm not opposed to adding several methods to River. However, we do like to keep things minimal, and not overwhelm users with many methods. This second method seems to be an extension of the first method. Maybe there's a way to have a single class with different parametrizations?

> Are there restrictions on the license?

You would have to abide by our BSD-3 license. I'm not sure your specific license encourages wide-spread adoption. But then again I'm not a licensing expert.

> Is there a guide for (potential) contributors?

[There is](https://github.com/online-ml/river/blob/main/CONTRIBUTING.md). It seems to be working fine, but let me know if you need help. In particular, you might want to make changes to the `drift` module.

How thing I see is that your code uses a lot of numpy. We try to avoid numpy because it's often slow for online updates. Do you think this will be a blocker? Or maybe you're confident numpy is the fastest option?
